Add multi-threading support

mrkarezina opened the pull request:
Added multithreading support with `batch_search`.

Tested to make sure output is the same. Environment: macOS 11.1 and Python 3.7.5

## Tests
**Original output**
```
python -m pyserini.search --topics robust04 --index robust04 --output run.robust04.txt --bm25
```
MD5 (run.robust04.txt) = d38797d0a284525d1d7ee35c00a9e5fd

```
python -m pyserini.search --topics msmarco_passage_dev_subset --index msmarco-passage --output runs/run.msmarco-passage.1.txt --bm25
```
MD5 (runs/run.msmarco-passage.1.txt) = 48ecace50020910580fc72d58fb9086c


**Modified script**
```
python -m pyserini.search --topics robust04 --index robust04 --output run.robust04-batch.1.txt --bm25
```
MD5 (run.robust04-batch.1.txt) = d38797d0a284525d1d7ee35c00a9e5fd

```
python -m pyserini.search --topics robust04 --index robust04 --output run.robust04-batch.2.txt --bm25 --threads 3 --batch-size 64
```
MD5 (run.robust04-batch.2.txt) = d38797d0a284525d1d7ee35c00a9e5fd


```
python -m pyserini.search --topics msmarco_passage_dev_subset --index msmarco-passage --output runs/run.msmarco-passage.1-batch.txt --bm25 --threads 4 --batch-size 64
```
MD5 (runs/run.msmarco-passage.1-batch.txt) = 48ecace50020910580fc72d58fb9086c

rodrigonogueira4 commented on pyserini/search/__main__.py:
```diff
@@ -131,36 +133,52 @@
+                results = searcher.batch_search(queries, ids, args.hits, args.threads)
+                results = sorted(results.items(), key=lambda item: int(item[0]))
+                batch_topics.clear()
+            else:
```
I don't think we need this `else`?

rodrigonogueira4 replied:
ah, nvm, we do need it!

rodrigonogueira4 reviewed (approved):
LGTM!

lintool commented:
The branches of the `if` statement are getting too unwieldy and should be broken up into separate functions, I think.

mrkarezina commented on pyserini/search/__main__.py:
```diff
@@ -131,36 +133,52 @@
-                if docid in unique_docs:
-                    continue
+                if use_prcl and len(hits) > (args.r + args.n):
+                    scores, docids = ranker.rerank(docids, scores)
 
                 if args.msmarco:
-                    target_file.write(f'{topic}\t{docid}\t{rank}\n')
+                    for i, docid in enumerate(docids):
+                        target_file.write(f'{topic}\t{docid}\t{i + 1}\n')
                 else:
-                    target_file.write(f'{topic} Q0 {docid} {rank} {hit.score:.6f} {tag}\n')
-                rank = rank + 1
-                unique_docs.add(docid)
-                if rank > args.max_passage_hits:
-                    break
+                    for i, (docid, score) in enumerate(zip(docids, scores)):
+                        target_file.write(f'{topic} Q0 {docid} {i + 1} {score:.6f} {tag}\n')
+            else:
+                unique_docs = set()
+                rank = 1
+                for hit in hits:
+                    docid, _ = hit.docid.split(args.max_passage_delimiter)
+                    if docid in unique_docs:
+                        continue
+
+                    if args.msmarco:
+                        target_file.write(f'{topic}\t{docid}\t{rank}\n')
+                    else:
+                        target_file.write(f'{topic} Q0 {docid} {rank} {hit.score:.6f} {tag}\n')
+                    rank = rank + 1
+                    unique_docs.add(docid)
+                    if rank > args.max_passage_hits:
+                        break
+        results.clear()
```
Should I refactor this into a seperate function? Which takes a "result", Tuple[str, List[JSimpleSearcherResult]], and writes it to the output file.

lintool replied:
I think the most important top-level separate is whether we're doing max-passage or not?

mrkarezina replied:
So best refactor it as two functions which write a "result", `write_result_max_passage` and `write_result`?

lintool replied:
I think so... give it a try and let's see how it looks?

mrkarezina commented on pyserini/search/__main__.py:
```diff
@@ -131,36 +133,52 @@
+            else:
+                unique_docs = set()
+                rank = 1
+                for hit in hits:
+                    docid, _ = hit.docid.split(args.max_passage_delimiter)
+                    if docid in unique_docs:
+                        continue
+
+                    if args.msmarco:
+                        target_file.write(f'{topic}\t{docid}\t{rank}\n')
+                    else:
+                        target_file.write(f'{topic} Q0 {docid} {rank} {hit.score:.6f} {tag}\n')
+                    rank = rank + 1
+                    unique_docs.add(docid)
+                    if rank > args.max_passage_hits:
+                        break
```
Or better refactor this into a function which writes hits to a file?

mrkarezina commented:
I refactored writing results into `write_result_max_passage` and `write_result`.

Tested the scripts as before and got the same checksums.

I’m not sure how to test the --max-passage flag. The doc id’s are just numbers, so when I add the flag there is an error on line 163 since there aren’t enough arguments to unpack. 
```
docid, _ = hit.docid.split(args.max_passage_delimiter)
```

lintool commented on pyserini/search/__main__.py:
```diff
@@ -130,37 +137,68 @@
+                continue
+
+        for result in results:
+            if not args.max_passage:
```
I think the if clause should be flipped... i.e., `if args.max_passage`... `else`

lintool commented:
I believe some of the integration tests depend on this also, so:

```
python -m unittest
python -m unittest discover -s integrations
```

Compare before and after, and make sure everything matches?

mrkarezina commented:
The integrations tests are failing, but the error seems to be unrelated. Both pyserini and anserini are in a common folder. Are you supposed to download these indexes in anserini? Currently the `indexes` folder in anserini is empty.

```
EE-------------------------
Running core17_bm25:
-------------------------
Index path '../anserini/indexes/lucene-index.core17.pos+docvectors+raw' does not exist or is not a directory.
Attempting to initialize pre-built index ../anserini/indexes/lucene-index.core17.pos+docvectors+raw.
Unrecognized index name ../anserini/indexes/lucene-index.core17.pos+docvectors+raw
E-------------------------
Running core17_bm25_rm3:
-------------------------
Index path '../anserini/indexes/lucene-index.core17.pos+docvectors+raw' does not exist or is not a directory.
Attempting to initialize pre-built index ../anserini/indexes/lucene-index.core17.pos+docvectors+raw.
Unrecognized index name ../anserini/indexes/lucene-index.core17.pos+docvectors+raw
```

Also I realized I didn't make a branch, so the pull bot will keep force pushing new commits to my repo.

lintool commented:
Oh, yea, for indexes that aren't there, okay to ignore.

mrkarezina commented:
Out of 20 tests, 18 failed. Errors consist of the one above and `FileNotFoundError: [Errno 2] No such file or directory: 'verify.anserini. ...` and `jnius.JavaException: JVM exception occurred: Java heap space` for `test_simple_fusion_search_integration`.

Not sure if the relevant ones passed.

lintool commented:
hrm, why does above msg say merged?

mrkarezina commented:
Not sure, maybe because i'm not on a branch?

lintool commented:
Odd...

Anyway, I'm off to bed now, I'll take a look at this tomorrow and run integration tests on your branch.

In the mean time, would you mind adding integration tests here:
https://github.com/castorini/pyserini/tree/master/integrations

Basically, take the tests you have

```
python -m pyserini.search --topics robust04 --index robust04 --output run.robust04.txt --bm25
python -m pyserini.search --topics msmarco_passage_dev_subset --index msmarco-passage --output runs/run.msmarco-passage.1.txt --bm25
```

Programmatically code them up. Run with different number of threads, e.g., 1 and 4, say - and confirm that the output is exactly the same (i.e., checksums match). That will futureproof.

lintool commented:
Ah, and for `--max-passage`, test this command:

```
python -m pyserini.search --topics msmarco_doc_dev --index msmarco-doc-per-passage --output runs/run.msmarco-doc.5.txt --bm25 --msmarco --hits 1000 --max-passage --max-passage-hits 100 &
```

Same thing, before and after. Integration testing with 1 vs. 4 threads.

(Sorry, I'm behind on documentation... so this hasn't been made public yet...)
